libs/providers/flagd-web/src/e2e/step-definitions/flag.ts

```diff
@@ -0,0 +1,352 @@
+    (expectedVariant: string, expectedReason: string) => {
+      expect(details).toBeDefined();
+      expect(details.variant).toEqual(expectedVariant);
+      expect(details.reason).toEqual(expectedReason);
+    },
+  );
+
+  then(/^the resolved string response should be "(.*)"$/, (expectedValue: string) => {
+    expect(value).toEqual(expectedValue);
+  });
+
+  when(
+    /^a non-existent string flag with key "(.*)" is evaluated with details and a default value "(.*)"$/,
+    async (key: string, defaultValue: string) => {
+      flagKey = key;
+      fallback = defaultValue;
+      details = await client.getStringDetails(flagKey, defaultValue);
+    },
+  );
+
+  then(/^the default string value should be returned$/, () => {
+    expect(details).toBeDefined();
+    expect(details.value).toEqual(fallback);
+  });
+
+  and(
+    /^the reason should indicate an error and the error code should indicate a missing flag with "(.*)"$/,
+    (errorCode: string) => {
+      expect(details).toBeDefined();
+      expect(details.reason).toEqual(StandardResolutionReasons.ERROR);
+      expect(details.errorCode).toEqual(errorCode);
+    },
+  );
+
+  when(
+    /^a string flag with key "(.*)" is evaluated as an integer, with details and a default value (\d+)$/,
+    async (key: string, defaultValue: string) => {
+      flagKey = key;
+      fallback = Number.parseInt(defaultValue);
+      details = await client.getNumberDetails(flagKey, Number.parseInt(defaultValue));
+    },
+  );
+
+  then(/^the default integer value should be returned$/, () => {
+    expect(details).toBeDefined();
+    expect(details.value).toEqual(fallback);
+  });
+
+  and(
+    /^the reason should indicate an error and the error code should indicate a type mismatch with "(.*)"$/,
+    (errorCode: string) => {
+      expect(details).toBeDefined();
+      expect(details.reason).toEqual(StandardResolutionReasons.ERROR);
+      expect(details.errorCode).toEqual(errorCode);
+    },
+  );
+
+  let ran: Promise<boolean>;
+  when('a PROVIDER_READY handler is added', () => {
+    ran = new Promise<boolean>((resolve) => {
+      client.addHandler(ProviderEvents.Ready, async () => {
+        resolve(true);
+      });
+    });
+  });
+  then('the PROVIDER_READY handler must run', () => {
+    expect(ran).toBeTruthy();
+  });
+
+  when('a PROVIDER_CONFIGURATION_CHANGED handler is added', () => {
+    ran = new Promise<boolean>((resolve) => {
+      client.addHandler(ProviderEvents.ConfigurationChanged, async (details) => {
+        // file writes are not atomic, so we get a few events in quick succession from the testbed
+        // some will not contain changes, this tolerates that; at least 1 should have our change
+
+        // All Flags are changed we do not provide a list of changed flags, that is the nature of the web sdk
+        //if (details?.flagsChanged?.length) {
+        //  flagsChanged = details?.flagsChanged;
+
+        resolve(true);
```

Comment:
This isn't testing anything.

Reply:
@aepfli 

@beeme1mr is right about this. `ran` will always be truthy b/c it's a promise.

You don't need to think in terms of promises here, you can do it simply with a var and some callbacks:

```ts
// define a var
let changeHandlerRan = false;

// add a handler
when('a PROVIDER_CONFIGURATION_CHANGED handler is added', () => {
  client.addHandler(ProviderEvents.ConfigurationChanged, () => {
    changeHandlerRan = true;
  });
});

// wait a bit (this is already implemented:
and(/^a flag with key "(.*)" is modified$/, async () => {
  // this happens every 1s in the associated container, so wait 3s
  await new Promise((resolve) => setTimeout(resolve, 3000));
});

// assert the handler ran (and mutated the var)
then('the PROVIDER_CONFIGURATION_CHANGED handler must run', async () => {
  expect(changeHandlerRan).toBeTruthy();
});
```

Re-writing these tests as above, the tests actually fail if you comment out the mutation of the `changeHandlerRan` var in the handler, proving the test actually works.

I think you'd have to do this for all the `PROVIDER_CONFIGURATION_CHANGED` and `PROVIDER_READY` event tests.

Reply:
changed it, and also added more proper documentation for the comments. i only need to create the issue to actually use the changed flags.